Corrected class name validation to no longer fail for Kotlin classes on class path containing special characters

studro opened the pull request:
Closes #1883.

----

Make sure these boxes are checked before submitting your PR -- thank you!

- [x] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

KengoTODA commented on spotbugs/src/main/java/edu/umd/cs/findbugs/util/ClassName.java:
```diff
@@ -175,9 +175,47 @@ public static String extractSimpleName(@DottedClassName String className) {
+    }
+
+    private static boolean isValidBinaryClassName(String className) {
+        return className.indexOf('.') == -1 &&
+                className.indexOf('[') == -1 &&
+                className.indexOf(';') == -1;
```
According to the [JVMS 4.2.2](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.2.2):

> An unqualified name must contain at least one Unicode code point and must not contain any of the ASCII characters . ; [ / (that is, period or semicolon or left square bracket or forward slash).

So better to check existence of `/` too.

studro replied:
I don't think this is entirely correct. An unqualified name may not contain `.`, `;`, `[`, or `/`, as you've quoted, but a binary class name consists of one or more unqualified names, which are separated by `/`s.

According to [JVMS 4.2.1](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.2.1): 

>  In this internal form, the ASCII periods (.) that normally separate the identifiers which make up the binary name are replaced by ASCII forward slashes (/). The identifiers themselves must be unqualified names (§4.2.2).

If we were to exclude `/` here, names of the form `com/foo/Example` would be rejected.

I suggest that this remains as is, but I am happy to take any further suggestions.

KengoTODA commented on spotbugs/src/main/java/edu/umd/cs/findbugs/util/ClassName.java:
```diff
@@ -175,9 +175,47 @@ public static String extractSimpleName(@DottedClassName String className) {
+                isValidClassFieldDescriptor(className);
+    }
+
+    private static boolean isValidBinaryClassName(String className) {
```
Better to have a hyperlink referring to the definition of 'valid binary class name':

```suggestion
    /**
     * @see <a href="https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.2.1">JVMS (Java SE 8 Edition) 4.2.2. Unqualified Names</a>
     */
    private static boolean isValidBinaryClassName(String className) {
```

Note that we may need to run `./gradlew spotlessApply` to reformat my snippet.

studro replied:
In your suggestion, the link is to section 4.2.1, but the link text is to 4.2.2. I think 4.2.1 is most correct, as it refers to class names, while 4.2.2 refers to the elements of the class name separated by `/` or `.`.

I also tried to wrap the link, as it resulted in a very long line.

Feel free to throw this back to me if you feel that 4.2.1 is incorrect as we should stick with 4.2.2 instead.

KengoTODA replied:
Thanks for your confirmation! The following sentence in the "4.2.1. Binary Class and Interface Names" is correct reference:

> The identifiers themselves must be unqualified names (§4.2.2).

KengoTODA commented on spotbugs/src/main/java/edu/umd/cs/findbugs/util/ClassName.java:
```diff
@@ -175,9 +175,47 @@ public static String extractSimpleName(@DottedClassName String className) {
+    }
+
+    private static boolean isValidArrayFieldDescriptor(String className) {
+        return className.startsWith("[") &&
+                (isValidArrayFieldDescriptor(className.substring(1))) ||
+                isValidClassFieldDescriptor(className.substring(1)) ||
+                isValidBaseTypeFieldDescriptor(className.substring(1));
```
It's just my personal and unreasonable motivation. JIT could be enough.

```suggestion
        String tail = className.substring(1);
        return className.startsWith("[") &&
                (isValidArrayFieldDescriptor(tail)) ||
                isValidClassFieldDescriptor(tail) ||
                isValidBaseTypeFieldDescriptor(tail);
```

studro replied:
I actually thinks this improves readability too, so this might be a good change. :+1:

However, I am concerned about the possibility of throwing an `IndexOutOfBoundsException` if this is called with an empty string (instead of returning false). I have added an `isEmpty()` check in the caller and documentation to mitigate this concern.

KengoTODA commented on spotbugs/src/main/java/edu/umd/cs/findbugs/util/ClassName.java:
```diff
@@ -175,9 +175,47 @@ public static String extractSimpleName(@DottedClassName String className) {
      */
     public static boolean isValidClassName(String className) {
-        // FIXME: should use a regex
+        return isValidBinaryClassName(className) ||
+                isValidDottedClassName(className) ||
+                isValidArrayFieldDescriptor(className) ||
+                isValidClassFieldDescriptor(className);
```
:+1: Confirmed that [all descriptions listed in the JVMS](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.3.2) are covered.

KengoTODA reviewed (changes requested):
Just one mandatory request exists in `isValidBinaryClassName()` methods, to follow the definition in the JVMS.
Other feedback are just my 2 cents. Thanks for your PR!

studro commented:
> Just one mandatory request exists in `isValidBinaryClassName()` methods, to follow the definition in the JVMS. Other feedback are just my 2 cents. Thanks for your PR!

Thanks for your quick and thorough review. I've attempted to address the comments you've raised and will be happy to make any further changes as necessary. 😄

studro commented on spotbugs/src/main/java/edu/umd/cs/findbugs/util/ClassName.java:
```diff
@@ -175,9 +175,64 @@ public static String extractSimpleName(@DottedClassName String className) {
+     * @param className a class name to test for validity - must be non-{@code null} and non-empty.
+     * @return {@code true} if {@code className} is a valid array field descriptor as
+     *          per JVMS 4.3.2, otherwise {@code false}
+     * @throws IndexOutOfBoundsException if {@code className} is {@code null} or empty.
```
I was thinking about the documentation of this later on in the day, and realised that I've made a mistake in documenting the exception behaviour. I think this minor change would be a good idea.

```suggestion
     * @throws IndexOutOfBoundsException if {@code className} is empty.
     * @throws NullPointerException if {@code className} is {@code null}.
```

studro commented:
I am very sorry to be difficult, but one final review is required, if possible.
I've updated the exception documentation as I noted in https://github.com/spotbugs/spotbugs/pull/1884#discussion_r772946651. While this was only a minor issue, what I had written was completely wrong.
I also noticed that I had the incorrect commit author email set for this project (my personal instead of no-reply github email), so I've updated history with a force push to correct that.
Thanks for your patience and hopefully this is now the end of this issue! 😊
